query_execution.go

```diff
@@ -0,0 +1,889 @@
+			r, err := executeBrambleStep(step)
+			if err != nil {
+				return nil, q.createGQLErrors(step, err)
+			}
+			results = append(results, *r)
+			continue
+		}
+
+		step := step
+		q.group.Go(func() error {
+			return q.executeRootStep(step)
+		})
+	}
+
+	readWg.Add(1)
+	go func() {
+		for result := range q.results {
+			results = append(results, result)
+		}
+		readWg.Done()
+	}()
+
+	if err := q.group.Wait(); err != nil {
+		return nil, gqlerror.List{
+			&gqlerror.Error{
+				Message: err.Error(),
+			},
+		}
+	}
+	close(q.results)
+	readWg.Wait()
+	return results, nil
+}
+
+func (q *queryExecution) executeRootStep(step *QueryPlanStep) error {
+	var document string
+
+	switch operationType := step.ParentType; operationType {
+	case queryObjectName, mutationObjectName:
+		document = strings.ToLower(operationType) + formatSelectionSet(q.ctx, q.schema, step.SelectionSet)
+	default:
+		return errors.New("expected mutation or query root step")
+	}
+
+	var data map[string]interface{}
+
+	err := q.executeDocument(document, step.ServiceURL, &data)
+	if err != nil {
+		q.writeExecutionResult(step, data, err)
+		return nil
+	}
+
+	q.writeExecutionResult(step, data, nil)
+
+	for _, childStep := range step.Then {
+		boundaryIDs, err := extractAndDedupeBoundaryIDs(data, childStep.InsertionPoint)
+		if err != nil {
+			return err
+		}
+		if len(boundaryIDs) == 0 {
+			continue
+		}
+
+		childStep := childStep
+		q.group.Go(func() error {
+			return q.executeChildStep(childStep, boundaryIDs)
+		})
+	}
+	return nil
+}
+
+func (q *queryExecution) executeDocument(document string, serviceURL string, response interface{}) error {
+	req := NewRequest(document).
+		WithHeaders(GetOutgoingRequestHeadersFromContext(q.ctx))
+	return q.graphqlClient.Request(q.ctx, serviceURL, req, &response)
+}
+
+func (q *queryExecution) writeExecutionResult(step *QueryPlanStep, data interface{}, err error) {
+	result := executionResult{
+		ServiceURL:     step.ServiceURL,
+		InsertionPoint: step.InsertionPoint,
+		Data:           data,
+	}
+	if err != nil {
+		result.Errors = q.createGQLErrors(step, err)
+	}
+
+	q.results <- result
+}
+
+func (q *queryExecution) executeChildStep(step *QueryPlanStep, boundaryIDs []string) error {
+	atomic.AddInt32(&q.requestCount, 1)
+	if q.requestCount > q.maxRequest {
+		return fmt.Errorf("exceeded max requests of %v", q.maxRequest)
+	}
+
+	boundaryField := q.boundaryFields.Field(step.ServiceURL, step.ParentType)
+
+	documents, err := buildBoundaryQueryDocuments(q.ctx, q.schema, step, boundaryIDs, boundaryField, 50)
+	if err != nil {
+		return err
+	}
+
+	data, err := q.executeBoundaryQuery(documents, step.ServiceURL, boundaryField)
+	if err != nil {
+		q.writeExecutionResult(step, data, err)
+		return nil
+	}
+
+	q.writeExecutionResult(step, data, nil)
+
+	if len(data) > 0 {
+		for _, childStep := range step.Then {
+			boundaryResultInsertionPoint, err := trimInsertionPointForNestedBoundaryStep(data, childStep.InsertionPoint)
```

Comment:
@lucianjon Instead of having the execution code do the trimming here, would it be possible for the planning code to provide the correct insertion point in the first place?

Reply:
As discussed offline, the insertion point is still correct in the context of merging the step's result into the overall response data. This case here is an exception as the result of a boundary query only contains a small piece of the entire response.

The previous code got around this as it shared a single map to store results in and would merge as it went, which meant the insertion point would always be a valid path inside the response data passed to a child step.